### DISCUSSION TITLE
Multiple concurrent BMP devices?

### DISCUSSION THREAD

**Discussion by jnz86** [blackmagic-debug/blackmagic#1929] [Q&A]
Should it be possible to address, say eight, probes at once? All in different states of flashing devices? 

I want to make a gang-of-eight / octopus programmer. I was thinking a raspPi, linux, eight BMP v2.3s. Then with python, run eight instances of the flashing program. At each port, detect, flash, wait for a new module, repeat. 

The things I am not clear on before I start research on this are:

* Should there be an issue addressing eight at once? I won't be able to use PID as they will all be the same. I assume I can read a serial number of a BMP?
* Is there anything in the driver that would prevent concurrent use? Like getting some single semaphore item from the OS?

I think those are two main issues for now. Thanks!

**Comment by dragonmux** [comment]
Each BMP is entirely independent of all others, and just shows up as different USB devices on the host. If you run the [project's udev rules](https://github.com/blackmagic-debug/blackmagic/tree/main/driver) then you can refer to each one by its serial number as in `/dev/ttyBmpGdb8BB20695` (for an example).

You can use the GDB remote protocol to connect into 8 different probes and control them separately just fine as there will be 8 `/dev/ttyBmpGdb[serial]` devices, all independent. One GDB can't talk to 8 simultaneously as that's not how GDB works - it can only connect to one remote at a time and must remain connected during the whole `load` process; but that's not a limitation if talking the serial protocol from a Python script.. just wrap that up in a class or function and call that function with each of the 8 file handles concurrently.
